<a href="https://colab.research.google.com/github/annbrennan/annbrennan.github.io/blob/main/spotify_willie_nelson.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Spotify Data Frame

Below is the **Spotify Data Frame** that reads the file spotify_all.csv containing data of Spotify users’ playlist information (Source: Spotify Million Playlist Dataset Challenge).

In [4]:
import pandas as pd
from google.colab import data_table
data_table.enable_dataframe_formatter()

spotify = pd.read_csv('https://bcdanl.github.io/data/spotify_all.csv')
spotify

pid playlist_name  pos        artist_name  \
0            0    Throwbacks    0      Missy Elliott   
1            0    Throwbacks    1     Britney Spears   
2            0    Throwbacks    2            Beyoncé   
3            0    Throwbacks    3  Justin Timberlake   
4            0    Throwbacks    4             Shaggy   
...        ...           ...  ...                ...   
198000  999998            ✝️    6       Chris Tomlin   
198001  999998            ✝️    7       Chris Tomlin   
198002  999998            ✝️    8            Crowder   
198003  999998            ✝️    9       Chris Tomlin   
198004  999998            ✝️   10   Caleb and Kelsey   

                                        track_name  duration_ms  \
0       Lose Control (feat. Ciara & Fat Man Scoop)       226863   
1                                            Toxic       198800   
2                                    Crazy In Love       235933   
3                                   Rock Your Body       267266   
4                                     It Wasn't Me       227600   
...                                            ...          ...   
198000                                   Waterfall       209573   
198001                                    The Roar       220106   
198002        Lift Your Head Weary Sinner (Chains)       224666   
198003                                We Fall Down       280960   
198004      10,000 Reasons / What a Beautiful Name       178189   

                                            album_name  
0                                         The Cookbook  
1                                          In The Zone  
2         Dangerously In Love (Alben für die Ewigkeit)  
3                                            Justified  
4                                             Hot Shot  
...                                                ...  
198000                                    Love Ran Red  
198001                                    Love Ran Red  
198002                                    Neon Steeple  
198003  How Great Is Our God: The Essential Collection  
198004          10,000 Reasons / What a Beautiful Name  

[198005 rows x 7 columns]

## Variable Description

* **pid:** playlist ID; unique ID for playlist
* **playlist_name:** a name of playlist
* **pos:** a position of the track within a playlist (starting from 0)
* **artist_name:** name of the track’s primary artist
* **track_name:** name of the track
* **duration_ms:** duration of the track in milliseconds
* **album_name:** name of the track’s album

## Favorite Artist

My favorite artist in the Spotify data frame is **Willie Nelson**.

All of Willie Nelson's songs in the Spotify data frame and their details are displayed below:

In [6]:
(
    spotify
    .set_index('artist_name')
    .loc['Willie Nelson']
)

pid playlist_name  pos  \
artist_name                                
Willie Nelson       9   old country    0   
Willie Nelson       9   old country    1   
Willie Nelson      83       Florida   66   
Willie Nelson      83       Florida   67   
Willie Nelson      90  For the Road   67   
...               ...           ...  ...   
Willie Nelson  999873          dads    2   
Willie Nelson  999873          dads    6   
Willie Nelson  999873          dads   45   
Willie Nelson  999897         Kings  153   
Willie Nelson  999936     Road Trip   67   

                                            track_name  duration_ms  \
artist_name                                                           
Willie Nelson                               Highwayman       182973   
Willie Nelson                               Highwayman       182973   
Willie Nelson                        Always On My Mind       212666   
Willie Nelson                         Luckenback Texas        94520   
Willie Nelson                               Highwayman       182973   
...                                                ...          ...   
Willie Nelson                        I Gotta Get Drunk       129759   
Willie Nelson                         Pancho and Lefty       286066   
Willie Nelson  Seven Spanish Angels (With Ray Charles)       229533   
Willie Nelson                        On the Road Again       153106   
Willie Nelson                        On the Road Again       153106   

                                        album_name  
artist_name                                         
Willie Nelson                      Nashville Rebel  
Willie Nelson                      Nashville Rebel  
Willie Nelson                    Always On My Mind  
Willie Nelson                      Country Outlaws  
Willie Nelson                      Nashville Rebel  
...                                            ...  
Willie Nelson                     Walking the Line  
Willie Nelson                     Walking the Line  
Willie Nelson                          Half Nelson  
Willie Nelson  Greatest Hits (& Some That Will Be)  
Willie Nelson  Greatest Hits (& Some That Will Be)  

[62 rows x 6 columns]

## Number of Songs

In [7]:
willie_nelson = (
    spotify
    .set_index('artist_name')
    .loc['Willie Nelson']
)
len(willie_nelson)

62

There are 62 Willie Nelson songs in this data frame.

## Longest Song

In [8]:
(
    willie_nelson
    .sort_values(['duration_ms'], ascending = False)
    .head(1)
)

pid playlist_name  pos                  track_name  \
artist_name                                                         
Willie Nelson  114        Willie   13  Sunday Mornin' Comin' Down   

               duration_ms                         album_name  
artist_name                                                    
Willie Nelson       422173  Willie Nelson Sings Kristofferson

The longest Willie Nelson song in this data set is **"Sunday Mornin' Comin' Down"** at 422173 milliseconds.

## Shortest Song

In [9]:
(
    willie_nelson
    .sort_values(['duration_ms'], ascending = True)
    .head(1)
)

pid playlist_name  pos        track_name  duration_ms  \
artist_name                                                            
Willie Nelson   83       Florida   67  Luckenback Texas        94520   

                    album_name  
artist_name                     
Willie Nelson  Country Outlaws

The shortest Willie Nelson song is **"Luckenback Texas"** at 94520 milliseconds.

## Songs between 150000 and 200000 ms in the "Willie" playlist

Below are all songs in the "Willie" playlist that are greater than 150000 ms but less than 200000 ms:

In [10]:
greater_than_15 = willie_nelson['duration_ms'] > 150000
less_than_20 = willie_nelson['duration_ms'] < 200000
willie_playlist = willie_nelson['playlist_name'] == 'Willie'
willie_nelson[greater_than_15 & less_than_20 & willie_playlist]

pid playlist_name  pos  \
artist_name                             
Willie Nelson  114        Willie    1   
Willie Nelson  114        Willie    3   
Willie Nelson  114        Willie    8   
Willie Nelson  114        Willie   10   
Willie Nelson  114        Willie   11   
Willie Nelson  114        Willie   12   
Willie Nelson  114        Willie   19   
Willie Nelson  114        Willie   21   
Willie Nelson  114        Willie   22   

                                                track_name  duration_ms  \
artist_name                                                               
Willie Nelson                     Little House on the Hill       182080   
Willie Nelson                Nothing I Can Do About It Now       199160   
Willie Nelson                                   Night Life       197813   
Willie Nelson                  If You Can Touch Her At All       183866   
Willie Nelson        Stay All Night (Stay A Little Longer)       154160   
Willie Nelson                        Sad Songs And Waltzes       185213   
Willie Nelson                           Fly Me To The Moon       169960   
Willie Nelson                                    Mona Lisa       151826   
Willie Nelson  Texas On A Saturday Night (With Mel Tillis)       159533   

                                               album_name  
artist_name                                                
Willie Nelson                         God's Problem Child  
Willie Nelson                        A Horse Called Music  
Willie Nelson  For the Good Times: A Tribute to Ray Price  
Willie Nelson     Funny How Time Slips Away - The Best Of  
Willie Nelson                              Shotgun Willie  
Willie Nelson                              Shotgun Willie  
Willie Nelson                            American Classic  
Willie Nelson                  Somewhere over the Rainbow  
Willie Nelson                                 Half Nelson

In [13]:
willie_songs = willie_nelson[greater_than_15 & less_than_20 & willie_playlist]
len(willie_songs)

9

There are 9 songs in the "Willie" playlist within this range.